### PR TITLE
Add in Subscription dataclass

### PR DESCRIPTION
**Pull request by wedamija**
This adds in the `Subscription` dataclass, which stores state about a subscription and implements `build_request`, which returns a `Request` that can be used to run a query.

We'll follow this up with further refactors to `parse_and_run_query` and related code to make it more able to cope with running subscription queries differently to api based queries (no rate limits, specific settings, etc).

### CODE REVIEW

File: snuba/state/rate_limit.py
Comment (wedamija) [inline snuba/state/rate_limit.py]: Btw does this even work? I'd assume we're going to hit the rest of this function once we exit the context manager.

Comment (tkaemming) [reply snuba/state/rate_limit.py]: 😩 

Nope. Good eye.

```python
>>> import contextlib
>>> @contextlib.contextmanager
... def ctx():
...   yield 1
...   yield 2
... 
>>> with ctx() as a:
...   print('hi')
... 
hi
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/usr/local/Cellar/python/3.7.2_2/Frameworks/Python.framework/Versions/3.7/lib/python3.7/contextlib.py", line 123, in __exit__
    raise RuntimeError("generator didn't stop")
RuntimeError: generator didn't stop
```

File: snuba/subscriptions/data.py
Comment (tkaemming) [inline snuba/subscriptions/data.py]: I think we'll end up with a subscription topic per dataset that is exclusively used for that dataset. If we do that, this value will be redundant. I'd also be a little hesitant about dynamically referencing datasets as part of `build_request` (though to be honest, I can't give you a good reason why.)

File: snuba/subscriptions/data.py
Comment (tkaemming) [inline snuba/subscriptions/data.py]: It might make sense to make these `timedelta` instances so we never have to have the "what unit is this again" conversation.

**Review by tkaemming — COMMENTED** [review]
Some initial thoughts on the data structure attributes themselves inline — I'll take a closer look at the query running aspects noted in the description and in `build_request` tomorrow.

File: snuba/subscriptions/data.py
Comment (fpacifici) [inline snuba/subscriptions/data.py]: Would it be worth parsing the query and store a Query object in the subscription instead of another partial representation of the query (conditions + aggregations).
It would require more refactoring of the validation but then we would not need to re-parse it every time.

**Review by fpacifici — COMMENTED** [review]
About the rate limiting. Right now it happens very low in the stack right before sending the query to clickhouse (kind of makes sense because it has to happen after a few other steps like returning cached responses). 
We are already controlling it through QuerySettings, which are partially controlled by the dataset, thus I would build on top of this to skip that phase by retaining everything else in the query processing.

**Comment by tkaemming** [comment]
> Currently this is tested using `parse_and_run_query`. This works fine at the moment as a test, but in production we don't want subscription queries to interfere with api queries, and vice versa. I don't know whether we want to rate limit subscription queries at all at this point, but we definitely want them to be using a different set of buckets to api queries if they do.
> 
> So we have two main directions we could take this:
> 
> * We could refactor `parse_and_run_query` so that we can run a query and just skip rate limiting, max_threads, etc.
> * We could override the request settings to for subscription queries so that we return no rate limits, then modify the ratelimiter to skip all the calls to redis if no rate limits are set.

Ok, some thoughts on this broader approach now. `parse_and_run_query` does several mutations or other stateful things, and I think it's helpful to enumerate them for clarity so you know what they are, so I remember what we do, and others can voice their opinions if they disagree with my conclusions. I'm mostly focusing on the behavior that I'm aware of for the events dataset, I am less familiar with the transactions implementation.

- **"Splits" the query.** (This is [applied as a decorator to `parse_and_run_query`](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/views.py#L304) and is easy to miss if you're not looking for it.) There are two types of splitting, [time splitting](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/split.py#L49-L60) and [column splitting](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/split.py#L137-L145). I don't think either of these apply here, as we should always have aggregations ([checked here](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/split.py#L40)) and will not be sorting by timestamp ([checked here](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/split.py#L44)).
- **Applies [dataset-specific extensions](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/views.py#L310-L314).** I do think this is necessary for subscriptions. The `ProjectExtension` needs to be applied to queries to the events dataset ([it is specified here](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/datasets/events.py#L384-L386)) because the [`ProjectWithGroupsProcessor`](https://github.com/getsentry/snuba/blob/94092d9bbc35c7f9b6f50acfa6a76d546cfd993a/snuba/query/project_extension.py#L91-L119) _must_ mutate the query to exclude groups that have been deleted or flip the final flag on if there have been too many groups deleted. (This broader behavior is [explained in this blog post](https://blog.sentry.io/2019/10/25/how-to-mutate-data-in-a-system-designed-for-immutable-data).)
- **Applies [dataset default conditions](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/views.py#L316).** I don't think this is strictly necessary as-is because [this method in the base dataset class is never overridden](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/datasets/dataset.py#L76-L81), but we should probably keep it, or remove it from the Dataset interface.
- **Handles ["turbo" mode](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/views.py#L318-L319).** We shouldn't need this. If we need to run these queries in turbo, something is wrong. This could return inaccurate results.
- **Applies [query processors](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/views.py#L321-L322).** We do need this, as it adds the [`PrewhereProcessor`](https://github.com/getsentry/snuba/blob/9bdcf8f49b50a19ca1fbabdc8d190bb2b07a3b8a/snuba/datasets/events.py#L394-L395), which is one of our main optimization paths.
- **Adds [mandatory conditions specified by the data source](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/views.py#L324-L325).** We need this. This adds the [not deleted condition](https://github.com/getsentry/snuba/blob/94092d9bbc35c7f9b6f50acfa6a76d546cfd993a/snuba/datasets/events.py#L243) to all queries.
- It then [calls out to `raw_query`](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/views.py#L345), which:
  - **Allows [bypassing the uncompressed cache](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/query.py#L57-L61).** This largely depends on the complexity of the conditions. This is probably useful.
  - **Applies [deduplication](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/query.py#L67).** This is an opinion and not necessarily fact, but I don't think this will be particularly useful here.
  - **Checks the [cache for this query result](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/query.py#L70).** Another opinion, but I don't think this will be particularly useful.
  - **Applies [rate limits](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/query.py#L86-L88).** This is probably the biggest question, like you mentioned. I don't think rate limits are particularly beneficial here: the number of concurrent queries is bounded by the number of subscription workers (consumers) we run here, and the number of concurrent queries per worker (size of the thread pool.) I also don't like the idea of the web requests slowing down subscriptions, or vice versa. I think these pools should be considered as distinct as possible. This is probably more of an operational concern than anything. We should get more opinions on this.
  - **Changes [`max_threads` based on the rate limits](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/query.py#L96-L104).** I don't think this will be helpful, since these queries have to run in consistent mode. (See below.)
  - **Applies [consistency settings](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/query.py#L106-L113).** These queries must run in consistent mode, for the same reason alert queries do. (This is [described somewhat in this blog post](https://blog.sentry.io/2019/09/17/how-to-get-stronger-consistency-out-of-a-datastore/), we can provide more detail about the specific relationship to alerts if needed.)
  - **[Runs the query against the database.](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/query.py#L116-L123)** We have to do this. Duh.
  - **Updates [the cache with the query result](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/api/query.py#L135-L137).** Whether or not this happens depends on the caching question above.
  - There's also a lot of metrics and HTTP-specific stuff thrown into this method. Some of that is probably useful, some of it probably isn't. We'll need to take a closer look at that, but that's more implementation details than anything that could change the broader approach here.

**Comment by fpacifici** [comment]
_I am less familiar with the transactions implementation._
All the above apply. What can make things different are Groups (joins) and Discover, but as long as you do the same steps you mentioned as "we should do" then we are fine.

About the query processors, these should be considered mandatory beyond PREWHERE since most query processing will happen there very soon.

File: snuba/subscriptions/data.py
Comment (wedamija) [reply snuba/subscriptions/data.py]: This seems like it'd be a good idea. If we do this, are there any risks involved with the `Query` object changing over time?

**Comment by tkaemming** [comment]
Ok, so following up on the previous comment with some opinions. Many of these don't really affect this change itself, but are my thoughts about how this change will eventually integrate with the broader system architecture.

- **Despite my earlier comments, I do think that `parse_and_run_query` is probably the right level of abstraction for running these queries.** I think it's more valuable to have all queries/requests go through a single execution function rather than optimizing the external and internal paths independently. These two paths share more functionality than they don't.
- **We need a different set of settings for different execution environments.** We need to be able to turn off caching and/or deduplication for subscriptions, and leave it on for the HTTP API. Reinforcing the previous point: this is configuration distinction, not a code path or functional one.
- **Since using `parse_and_run_query` for subscriptions will be outside of the HTTP request and response cycle, we should move it from `snuba/views.py` and put it somewhere that reflects it's more general use at this point.** When we do that, we should also remove some of the coupling between this function return values and HTTP statuses and responses. Things like rate limited errors should throw exceptions, not early return with an HTTP status code and response body. We should catch these exceptions and return the appropriate status code with contextual information in the body instead. [Some of this work has already been done](https://github.com/getsentry/snuba/blob/abc944a3eed660024b13737da23822dfc1a4a96f/snuba/views.py#L99-L138) but we need to make sure it continues.
- **I still don't think rate limiting makes sense here.** How would the subscription consumer handle rate limits? Would it forward the rate limited error to the result channel? I think that would be a bad idea. I guess the other option would be to just to spin loop on queries that are blocked on rate limits until they succeed. If there are project-specific rate limits, that means a single project could block the consumer until it's rate limit clears. We'll need to keep an eye on usage (probably in terms of the number of subscriptions, how frequent those subscriptions are, and how many columns are referenced in the query and over how long of a time range) but I don't think rate limiting is after the subscriptions have been created is going to the right way to maintain quality of service.
- **We're going to need to eventually be able to apply extensions to multiple queries/requests at once.** This would allow us to optimize things like the [`ProjectWithGroupsProcessor`](https://github.com/getsentry/snuba/blob/94092d9bbc35c7f9b6f50acfa6a76d546cfd993a/snuba/query/project_extension.py#L91-L119) by enabling it to get the exclusion groups in parallel. We can start off doing this in serial for the early version(s), and potentially eventually do it in a thread pool to parallelize it somewhat, but I'd think it's more efficient to do this as a Redis pipeline from one client rather than N clients (one for each thread 
in the executor thread pool.)
- **We will always need to force turbo off, and consistent on.** I think this implies that setting the request settings on subscriptions should always be an internal thing.

File: snuba/subscriptions/data.py
Comment (tkaemming) [reply snuba/subscriptions/data.py]: I like this general idea, but have a few questions/concerns with that approach:

1. The `Query` instance (I'm assuming we're talking about [this one](https://github.com/getsentry/snuba/blob/f6567d23c9295614c06c567eea0a12eb3c81f298/snuba/query/query.py#L49)) is already bound to a data source. That seems problematic for serialization, schema migrations, etc. [This comment from below](https://github.com/getsentry/snuba/pull/618/files#r354089337) would also apply here. Changing so that a query isn't bound to a dataset or data source (table, optionally with joins) would alleviate these concerns, but that's a pretty wide reaching change. At a high level, I think `source.execute(query)` makes a lot more sense to me than `query.execute()` with the source already bound, for what it's worth.
2. The query contains nodes that are variables (current timestamp.) I'm not sure how we'd express those. Adding a `Variable` node to the AST would enable that, but obviously opens another complexity can of worms.

Comment (wedamija) [reply snuba/subscriptions/data.py]: Also if we're planning on using `parse_and_run_query` then it might make sense to keep things in the same format.

File: snuba/subscriptions/data.py
Comment (tkaemming) [inline snuba/subscriptions/data.py]: Note: this implies the dataset uses the [`TimeSeriesExtensionProcessor`](https://github.com/getsentry/snuba/blob/7023fed81d1981ece9ea7ef745450deb5fec3345/snuba/query/timeseries.py#L14), which is probably an accurate and good thing — we should make sure to enforce that you can only create subscriptions to datasets that support this functionality as part of the subscription creation APIs.

File: snuba/subscriptions/data.py
Comment (wedamija) [reply snuba/subscriptions/data.py]: One thought here is that the subscription is to a query, and dataset is a core part of a query, since the conditions that can be used are tied to that. I can see that it might not make sense for `build_request` to return it, since it overloads the name of the method.

I guess one concern could be if dataset names change, but I'm not sure whether that's likely?

Overall I don't feel too strongly about keeping it here though.

**Comment by wedamija** [comment]
So my thoughts here are that it probably makes sense to get this PR tidied up and then merged, and  then work on making changes to `parse_and_run_query` and query running in general in separate PRs so that we can keep the size of changes manageable. I'm going to go ahead with that in a separate branch, if anyone disagrees I can just merge the changes back in.

File: snuba/subscriptions/data.py
Comment (wedamija) [reply snuba/subscriptions/data.py]: For the moment it seems like it makes sense to keep this as is, so that we don't blow out the complexity.

**Review by tkaemming — CHANGES_REQUESTED** [review]
A few minor things in here but overall I think we can move forward with this and make improvements later like you suggested.

File: snuba/subscriptions/data.py
Comment (wedamija) [reply snuba/subscriptions/data.py]: Ahh, I looked for this in snuba/query/types.py and assumed there wasn't a constant defined for it.

File: snuba/subscriptions/data.py
Comment (tkaemming) [reply snuba/subscriptions/data.py]: Oops, yeah — you're right. The ID also needs to include the partition key so that operations after create can be routed to the correct project. This might tie in a bit with this comment then too: https://github.com/getsentry/snuba/pull/618#discussion_r356214864

File: snuba/views.py
Comment (tkaemming) [reply snuba/views.py]: Makes sense, let's do it here if that can be done cleanly, otherwise it's fine to move forward as-is.

File: snuba/subscriptions/data.py
Comment (fpacifici) [reply snuba/subscriptions/data.py]: ok, no, if this represent a long lived subscription then, definitively DO NOT use Query. 
That is a throw away object to execute the query that gets updated in place multiple times every time the query is executed. I thought this would represent something different. That represents one instance of a query.

File: snuba/subscriptions/data.py
Comment (fpacifici) [inline snuba/subscriptions/data.py]: why isn't the dataset part of the subscription and instead passed at the moment the request is built ? I mean conceptually. Is it a serialization problem ? Specifically who is going to provide it  and how do you plan disambiguate multiple subscription designed for different datasets ?

Comment (wedamija) [reply snuba/subscriptions/data.py]: https://github.com/getsentry/snuba/pull/618#discussion_r354089337

I don't feel strongly either way about this.

File: snuba/subscriptions/data.py
Comment (tkaemming) [inline snuba/subscriptions/data.py]:
```suggestion
        Returns a Request that can be used to run a query via `parse_and_run_query`.
```

**Review by tkaemming — APPROVED** [review]
This seems like a good start. I think we'll figure out what to do with `id` and `project_id` as we get further along.

**Pull request merged**
